kmath-commons-rng-part/src/commonMain/kotlin/scientifik/commons/rng/sampling/distribution/LargenMeanPoissonSampler.kt

```diff
@@ -0,0 +1,233 @@
+package scientifik.commons.rng.sampling.distribution
+
+import scientifik.commons.rng.UniformRandomProvider
+import kotlin.math.*
+
+class LargeMeanPoissonSampler : SharedStateDiscreteSampler {
+    private val rng: UniformRandomProvider
+    private val exponential: SharedStateContinuousSampler
+    private val gaussian: SharedStateContinuousSampler
+    private val factorialLog: InternalUtils.FactorialLog
+    private val lambda: Double
+    private val logLambda: Double
+    private val logLambdaFactorial: Double
+    private val delta: Double
+    private val halfDelta: Double
+    private val twolpd: Double
+    private val p1: Double
+    private val p2: Double
+    private val c1: Double
+    private val smallMeanPoissonSampler: SharedStateDiscreteSampler
+
+    constructor(
+        rng: UniformRandomProvider,
+        mean: Double
+    ) {
+        require(mean >= 1) { "mean is not >= 1: $mean" }
+        // The algorithm is not valid if Math.floor(mean) is not an integer.
+        require(mean <= MAX_MEAN) { "mean $mean > $MAX_MEAN" }
+        this.rng = rng
+        gaussian = ZigguratNormalizedGaussianSampler(rng)
+        exponential = AhrensDieterExponentialSampler.of(rng, 1.0)
+        // Plain constructor uses the uncached function.
+        factorialLog = NO_CACHE_FACTORIAL_LOG!!
+        // Cache values used in the algorithm
+        lambda = floor(mean)
+        logLambda = ln(lambda)
+        logLambdaFactorial = getFactorialLog(lambda.toInt())
+        delta = sqrt(lambda * ln(32 * lambda / PI + 1))
+        halfDelta = delta / 2
+        twolpd = 2 * lambda + delta
+        c1 = 1 / (8 * lambda)
+        val a1: Double = sqrt(PI * twolpd) * exp(c1)
+        val a2: Double = twolpd / delta * exp(-delta * (1 + delta) / twolpd)
+        val aSum = a1 + a2 + 1
+        p1 = a1 / aSum
+        p2 = a2 / aSum
+
+        // The algorithm requires a Poisson sample from the remaining lambda fraction.
+        val lambdaFractional = mean - lambda
+        smallMeanPoissonSampler =
+            if (lambdaFractional < Double.MIN_VALUE) NO_SMALL_MEAN_POISSON_SAMPLER else  // Not used.
+                KempSmallMeanPoissonSampler.of(rng, lambdaFractional)
+    }
+
+    internal constructor(
```

Comment:
Probably should be refactored to use companion functions instead of secondary constructors.